File: apps/api/app/services/source_url.py
"""Validation and detection for supported review-platform URLs."""

import re
from urllib.parse import unquote, urlparse

from app.domain import ReviewPlatform
from app.errors import AppValidationError
from app.schemas import ApiErrorDetail

_GOOGLE_HOST = re.compile(r"^(www\.|maps\.)?google(\.[a-z]{2,3}){1,2}$")
_GOOGLE_SHORT_HOSTS = {"maps.app.goo.gl", "goo.gl", "g.co"}
_AMAZON_HOSTS = {"amazon.com", "www.amazon.com"}
_AMAZON_ASIN = re.compile(r"/(?:dp|gp/product)/([A-Z0-9]{10})(?:[/?]|$)", re.IGNORECASE)
_AMAZON_SLUG_ASIN = re.compile(r"/[^/]+/dp/([A-Z0-9]{10})(?:[/?]|$)", re.IGNORECASE)
_PLACE_SEGMENT = re.compile(r"/place/([^/@]+)")


def _reject(issue: str) -> None:
    raise AppValidationError(
        "The review source is not valid",
        [ApiErrorDetail(field="source_url", issue=issue)],
    )


def validate_source_url(raw_url: str) -> str:
    """Return the normalized source URL, or raise AppValidationError."""
    candidate = (raw_url or "").strip()
    if not candidate:
        _reject("A review source URL is required.")

    try:
        parsed = urlparse(candidate)
    except ValueError:
        _reject("That does not look like a valid web address.")

    if parsed.scheme not in {"http", "https"}:
        _reject("The review source must be a http:// or https:// web address.")

    host = (parsed.hostname or "").lower()
    if not host:
        _reject("That does not look like a valid web address.")

    if host in _GOOGLE_SHORT_HOSTS:
        return candidate

    if host in _AMAZON_HOSTS:
        if not (
            _AMAZON_ASIN.search(parsed.path) or _AMAZON_SLUG_ASIN.search(parsed.path)
        ):
            _reject(
                "Paste an Amazon.com product link that contains a product ASIN, such as /dp/B012345678."
            )
        return candidate

    if not _GOOGLE_HOST.match(host):
        _reject(
            "ReviewLens supports Google Maps place links and Amazon.com product links."
        )

    if not parsed.path.startswith("/maps"):
        _reject(
            "That Google link is not a Google Maps place. Open the business on Google Maps and copy the link."
        )

    return candidate


def detect_platform(source_url: str) -> ReviewPlatform:
    """Identify the supported platform after validating its URL shape."""
    normalized_url = validate_source_url(source_url)
    host = (urlparse(normalized_url).hostname or "").lower()
    if host in _AMAZON_HOSTS:
        return ReviewPlatform.AMAZON
    return ReviewPlatform.GOOGLE_MAPS


def extract_amazon_asin(source_url: str) -> str | None:
    parsed = urlparse(source_url)
    if (parsed.hostname or "").lower() not in _AMAZON_HOSTS:
        return None
    match = _AMAZON_ASIN.search(parsed.path) or _AMAZON_SLUG_ASIN.search(parsed.path)
    return match.group(1).upper() if match else None


def derive_working_name(source_url: str) -> str | None:
    """Best-effort display name from a Maps place or Amazon product URL."""
    parsed = urlparse(source_url)
    host = (parsed.hostname or "").lower()
    if host in _AMAZON_HOSTS:
        asin = extract_amazon_asin(source_url)
        return f"Amazon Product {asin}" if asin else None

    match = _PLACE_SEGMENT.search(parsed.path)
    if not match:
        return None
    slug = unquote(match.group(1)).replace("+", " ").strip()
    return slug or None

File: apps/web/src/lib/sourceUrl.ts
/**
 * Client-side mirror of the backend's source rules. The server remains
 * authoritative; this exists only to give immediate feedback.
 */

const GOOGLE_HOST = /^(www\.|maps\.)?google(\.[a-z]{2,3}){1,2}$/
const GOOGLE_SHORT_HOSTS = ['maps.app.goo.gl', 'goo.gl', 'g.co']
const AMAZON_HOSTS = ['amazon.com', 'www.amazon.com']
const AMAZON_ASIN = /\/(?:dp|gp\/product)\/([A-Z0-9]{10})(?:[/?]|$)/i
const AMAZON_SLUG_ASIN = /\/[^/]+\/dp\/([A-Z0-9]{10})(?:[/?]|$)/i
const PLACE_SEGMENT = /\/place\/([^/@]+)/

export type SourcePlatform = 'google_maps' | 'amazon'

export function detectSourcePlatform(raw: string): SourcePlatform | null {
  if (validateSourceUrl(raw)) return null
  const parsed = new URL(raw.trim())
  const host = parsed.hostname.toLowerCase()
  if (AMAZON_HOSTS.includes(host)) return 'amazon'
  return 'google_maps'
}

export function validateSourceUrl(raw: string): string | null {
  const candidate = raw.trim()
  if (!candidate) return 'Paste a Google Maps place or Amazon.com product link to get started.'

  let parsed: URL
  try {
    parsed = new URL(candidate)
  } catch {
    return 'That does not look like a valid web address.'
  }

  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
    return 'The review source must be a http:// or https:// web address.'
  }

  const host = parsed.hostname.toLowerCase()
  if (GOOGLE_SHORT_HOSTS.includes(host)) return null

  if (AMAZON_HOSTS.includes(host)) {
    if (!(AMAZON_ASIN.test(parsed.pathname) || AMAZON_SLUG_ASIN.test(parsed.pathname))) {
      return 'Paste an Amazon.com product link that contains a product ASIN.'
    }
    return null
  }

  if (!GOOGLE_HOST.test(host)) {
    return 'ReviewLens supports Google Maps place links and Amazon.com product links.'
  }

  if (!parsed.pathname.startsWith('/maps')) {
    return 'That Google link is not a Google Maps place. Open the business on Google Maps and copy the link.'
  }

  return null
}

/** Working name derived from the place slug or product ASIN. */
export function deriveWorkingName(raw: string): string {
  try {
    const parsed = new URL(raw.trim())
    if (AMAZON_HOSTS.includes(parsed.hostname.toLowerCase())) {
      const match = AMAZON_ASIN.exec(parsed.pathname) ?? AMAZON_SLUG_ASIN.exec(parsed.pathname)
      if (match) return `Amazon Product ${match[1].toUpperCase()}`
    }
    const match = PLACE_SEGMENT.exec(parsed.pathname)
    if (match) {
      const slug = decodeURIComponent(match[1]).replace(/\+/g, ' ').trim()
      if (slug) return slug.slice(0, 200)
    }
  } catch {
    // Fall through to the placeholder name.
  }
  return 'Untitled Analysis'
}
